core/embed/rust/src/ui/layout/obj.rs

```diff
@@ -0,0 +1,350 @@
+}
+
+#[cfg(not(feature = "ui_debug"))]
+mod maybe_trace {
+    pub trait ObjComponentTrace: super::ObjComponent {}
+    impl<T> ObjComponentTrace for T where T: super::ObjComponent {}
+}
+
+/// Trait that combines `ObjComponent` with `Trace` if `ui_debug` is enabled,
+/// and pure `ObjComponent` otherwise.
+use maybe_trace::ObjComponentTrace;
+
+/// `LayoutObj` is a GC-allocated object exported to MicroPython, with type
+/// `LayoutObj::obj_type()`. It wraps a root component through the
+/// `ObjComponent` trait.
+#[repr(C)]
+pub struct LayoutObj {
+    base: ObjBase,
+    inner: RefCell<LayoutObjInner>,
+}
+
+struct LayoutObjInner {
+    root: Gc<dyn ObjComponentTrace>,
+    event_ctx: EventCtx,
+    timer_fn: Obj,
+}
+
+impl LayoutObj {
+    /// Create a new `LayoutObj`, wrapping a root component.
+    pub fn new(root: impl ObjComponentTrace + 'static) -> Result<Gc<Self>, Error> {
+        // SAFETY: We are coercing GC-allocated sized ptr into an unsized one.
+        let root =
+            unsafe { Gc::from_raw(Gc::into_raw(Gc::new(root)?) as *mut dyn ObjComponentTrace) };
+
+        Gc::new(Self {
+            base: Self::obj_type().as_base(),
+            inner: RefCell::new(LayoutObjInner {
+                root,
+                event_ctx: EventCtx::new(),
+                timer_fn: Obj::const_none(),
+            }),
+        })
+    }
+
+    /// Timer callback is expected to be a callable object of the following
+    /// form: `def timer(token: int, deadline_in_ms: int)`.
+    fn obj_set_timer_fn(&self, timer_fn: Obj) {
+        self.inner.borrow_mut().timer_fn = timer_fn;
+    }
+
+    /// Run an event pass over the component tree. After the traversal, any
+    /// pending timers are drained into `self.timer_callback`. Returns `Err`
+    /// in case the timer callback raises, `Ok` with the message otherwise.
+    fn obj_event(&self, event: Event) -> Result<Obj, Error> {
+        let inner = &mut *self.inner.borrow_mut();
+
+        // Clear the upwards-propagating paint request flag from the last event pass.
+        inner.event_ctx.clear_paint_requests();
+
+        // Send the event down the component tree. Bail out in case of failure.
+        // SAFETY: `inner.root` is unique because of the `inner.borrow_mut()`.
+        let msg = unsafe { Gc::as_mut(&mut inner.root) }.obj_event(&mut inner.event_ctx, event)?;
+
+        // All concerning `Child` wrappers should have already marked themselves for
+        // painting by now, and we're prepared for a paint pass.
+
+        // Drain any pending timers into the callback.
+        while let Some((token, deadline)) = inner.event_ctx.pop_timer() {
+            let token = token.try_into();
+            let deadline = deadline.try_into();
+            if let (Ok(token), Ok(deadline)) = (token, deadline) {
+                inner.timer_fn.call_with_n_args(&[token, deadline])?;
+            } else {
+                // Failed to convert token or deadline into `Obj`, skip.
```

Comment:
Would it make sense to panic here at least in debug mode, so that we catch such errors in CI?